sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/ProactiveConnectionManagementTest.java

```diff
@@ -0,0 +1,393 @@
+                    .flatMap(CosmosAsyncContainer::read)
+                    .flatMap(containerResponse -> rxDocumentClient
+                            .getPartitionKeyRangeCache()
+                            .tryGetOverlappingRangesAsync(
+                                    null,
+                                    containerResponse.getProperties().getResourceId(),
+                                    PartitionKeyInternalHelper.FullRange,
+                                    false,
+                                    null));
+
+            // 1. Extract all preferred read regions to proactively connect to.
+            // 2. Obtain partition addresses for a container for one read region, then mark that read region as unavailable.
+            // 3. This will force resolveAsync to use the next preferred read region in the next invocation.
+            // 4. This way we can verify that connections have been opened to all replicas across all proactive connection regions.
+            for (URI proactiveConnectionEndpoint : proactiveConnectionEndpoints) {
+                Flux.zip(asyncContainerFlux, partitionKeyRangeFlux)
+                        .flatMapIterable(containerToPartitionKeyRanges -> {
+                            List<ImmutablePair<PartitionKeyRange, CosmosAsyncContainer>> pkrToContainer = new ArrayList<>();
+                            for (PartitionKeyRange pkr : containerToPartitionKeyRanges.getT2().v) {
+                                pkrToContainer.add(new ImmutablePair<>(pkr, containerToPartitionKeyRanges.getT1()));
+                            }
+                            return pkrToContainer;
+                        })
+                        .flatMap(partitionKeyRangeToContainer -> {
+                            RxDocumentServiceRequest dummyRequest = RxDocumentServiceRequest.createFromName(
+                                    mockDiagnosticsClientContext(),
+                                    OperationType.Read,
+                                    partitionKeyRangeToContainer.getRight().getLink() + "/docId",
+                                    ResourceType.Document);
+                            dummyRequest.setPartitionKeyRangeIdentity(new PartitionKeyRangeIdentity(partitionKeyRangeToContainer.getLeft().getId()));
+                            return globalAddressResolver.resolveAsync(dummyRequest, false);
+                        })
+                        .delayElements(Duration.ofSeconds(3))
+                        .doOnNext(addressInformations -> {
+                            for (AddressInformation address : addressInformations) {
+                                endpoints.add(address.getPhysicalUri().getURI().getAuthority());
+                            }
+                        })
+                        .blockLast();
+
+                globalEndpointManager.markEndpointUnavailableForRead(proactiveConnectionEndpoint);
+            }
+
```

Comment:
Do we need to track somehow the time it takes to use the next read region to confirm it is not opening new connections on the fly (i.e. some way to verify it uses warmed up connections that have been pre-opened)?

Reply:
The unit test code wouldn't open connections - it is checking the status of addresses from cache. It makes certain assumptions of the internal implementation for validation - I agree, it would be good to also have an end-to-end test (including metrics and ideally some of the failure injections once available) in CI-pipeline - will use this issue to track it - https://github.com/Azure/azure-sdk-for-java/issues/33289